feat: create evm prover (#56)
celestiaorg/celestia-zkevm-ibc-demo: provers/evm-prover/README.md

# EVM Prover

The EVM Prover is a gRPC service that generates zero-knowledge proofs for EVM state transitions. It is designed to work with IBC (Inter-Blockchain Communication) and specifically implements proofs compatible with the ICS-07 Tendermint client specification.

## Usage

> [!WARNING] 
> This gRPC service is still under development and may not work as described

To run the server you will need to clone the repo and install rust and cargo. To run the node you also need to set the following environment variables:

- `TENDERMINT_RPC_URL` - the url of the tendermint chain you are proving.
- `RPC_URL` the json rpc url of the evm chain you are generating the proofs for.

> - `CONTRACT_ADDRESS` - the evm address of the tendermint sp1 ics07 contract.

Reviewer: nit: I think these environment variables don't reflect the environment variables expected by the EVM prover

Author: Updated.

To then run the server (on port `:50051`):

```
cargo run
```

> To use the SP1 Prover Network you should also populate the `SP1_PROVER` and `SP1_PRIVATE_KEY` environment variables. You can also use a `.env` file for all environment variables

Reviewer: question: should we provide an example.env that users can copy with pre-populated defaults for the environment variables?

Author: Added `.env.example`

# EVM Prover

The EVM Prover is a gRPC service that generates zero-knowledge proofs for EVM state transitions. It is designed to work with IBC (Inter-Blockchain Communication) and specifically implements proofs compatible with the ICS-07 Tendermint client specification.

## Usage

> [!WARNING]
> This gRPC service is still under development and may not work as described

To run the server you will need to clone the repo and install rust and cargo. To run the node you also need to set the following environment variables:

> - `CELESTIA_NODE_AUTH_TOKEN` - the auth token for the celestia node you are connecting to.

Reviewer 2: we have a node auth token file in test files so it should probably consume the key form there?

Author: yep

    cargo run
    ```

The `Info` endpoint returns the state transition and membership verification keys:

    ```shell
    grpcurl -plaintext localhost:50052 celestia.prover.v1.Prover/Info
    ```

> The `ProveStateTransition` endpoint invokes a state transition proof and returns an aggregate range proof:

Reviewer: [nit] idk if this is clearer

```suggestion
The `ProveStateTransition` endpoint generates a state transition proof for a range of EVM heights:
```

Author: Fixed.

    ```shell
    grpcurl -plaintext -d '{"client_id":"08-groth16-0"}' localhost:50052 celestia.prover.v1.Prover/ProveStateTransition

    {
    "proof": "...",
    "publicValues": "..."
    }
    ```

Note that this requires the IBC light clients to be setup first:

    ```shell
    make setup
    ```

The proving time for the ZK block range proof depends on the prover settings, i.e. prover network and mode.

> For reference in mock mode, the proof generation takes ~15 minutes for aggregating 10 blocks.

Reviewer: oof is there a way to accelerate this for development time? Can be a follow up.
